
# **Problem Statement:**
**Building a new Employee Security System**
You, as the Security analyst, at Stark Industries, have been tasked to build a new contactless employee check-in system. Currently the employees use a physical keycard for entry into the building like shown below. 

You have come up with a new idea that uses the employees smartphone and machine learning to provide a contactless system where when an employee enters the firm’s territory, his or her smartphone connects to the server and transmits data from the employee smartphone sensor data like the accelerometer's data. The server performs the calculations and determines this person as
one of the employees using Gait analysis. Essentially it compares the current pattern of the employee’s gait with the historial pattern and if there is a match, the doors automatically open for the employee to walk in.

To test your idea, you have built a dataset of 30 employees and their daily activities here. Design and develop a system that will perform the gait analysis. 

# **GAIT Analysis**

Gait analysis is an assessment of the way the body moves, usually by walking or running, from one place to another. In this problem statement using Gait analysis it determines person's movement like how he or she walk and compares the current pattern of the employee's with the help of historial pattern 

In [1]:
!unzip "/content/UCI HAR Dataset.zip"

'unzip' is not recognized as an internal or external command,
operable program or batch file.


### Logistic Regression

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [3]:
train_df = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt", delim_whitespace = True,names=list(range(0,561)))

train_df['Subject'] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\subject_train.txt", header = None, squeeze = True)

train_df["activity"] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\y_train.txt", header = None, squeeze = True)
# C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train

In [4]:
features = list()
with open(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [5]:
features

['tBodyAcc-mean()-X',
 'tBodyAcc-mean()-Y',
 'tBodyAcc-mean()-Z',
 'tBodyAcc-std()-X',
 'tBodyAcc-std()-Y',
 'tBodyAcc-std()-Z',
 'tBodyAcc-mad()-X',
 'tBodyAcc-mad()-Y',
 'tBodyAcc-mad()-Z',
 'tBodyAcc-max()-X',
 'tBodyAcc-max()-Y',
 'tBodyAcc-max()-Z',
 'tBodyAcc-min()-X',
 'tBodyAcc-min()-Y',
 'tBodyAcc-min()-Z',
 'tBodyAcc-sma()',
 'tBodyAcc-energy()-X',
 'tBodyAcc-energy()-Y',
 'tBodyAcc-energy()-Z',
 'tBodyAcc-iqr()-X',
 'tBodyAcc-iqr()-Y',
 'tBodyAcc-iqr()-Z',
 'tBodyAcc-entropy()-X',
 'tBodyAcc-entropy()-Y',
 'tBodyAcc-entropy()-Z',
 'tBodyAcc-arCoeff()-X,1',
 'tBodyAcc-arCoeff()-X,2',
 'tBodyAcc-arCoeff()-X,3',
 'tBodyAcc-arCoeff()-X,4',
 'tBodyAcc-arCoeff()-Y,1',
 'tBodyAcc-arCoeff()-Y,2',
 'tBodyAcc-arCoeff()-Y,3',
 'tBodyAcc-arCoeff()-Y,4',
 'tBodyAcc-arCoeff()-Z,1',
 'tBodyAcc-arCoeff()-Z,2',
 'tBodyAcc-arCoeff()-Z,3',
 'tBodyAcc-arCoeff()-Z,4',
 'tBodyAcc-correlation()-X,Y',
 'tBodyAcc-correlation()-X,Z',
 'tBodyAcc-correlation()-Y,Z',
 'tGravityAcc-mean()-X',
 'tGravityA

In [6]:
test_df = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\X_test.txt", delim_whitespace = True,names=list(range(0,561)))

test_df['Subject'] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\subject_test.txt", header = None, squeeze = True)

test_df["activity"] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\test\y_test.txt", header = None, squeeze = True)

In [7]:
train_df.to_csv('train.csv',index_label=False)

In [8]:
test_df.to_csv('test.csv',index_label=False)

In [9]:
train_df= pd.read_csv('train.csv')

In [10]:
train_df.head()

0         1         2         3         4         5         6  \
0  0.288585 -0.020294 -0.132905 -0.995279 -0.983111 -0.913526 -0.995112   
1  0.278419 -0.016411 -0.123520 -0.998245 -0.975300 -0.960322 -0.998807   
2  0.279653 -0.019467 -0.113462 -0.995380 -0.967187 -0.978944 -0.996520   
3  0.279174 -0.026201 -0.123283 -0.996091 -0.983403 -0.990675 -0.997099   
4  0.276629 -0.016570 -0.115362 -0.998139 -0.980817 -0.990482 -0.998321   

          7         8         9  ...       553       554       555       556  \
0 -0.983185 -0.923527 -0.934724  ... -0.710304 -0.112754  0.030400 -0.464761   
1 -0.974914 -0.957686 -0.943068  ... -0.861499  0.053477 -0.007435 -0.732626   
2 -0.963668 -0.977469 -0.938692  ... -0.760104 -0.118559  0.177899  0.100699   
3 -0.982750 -0.989302 -0.938692  ... -0.482845 -0.036788 -0.012892  0.640011   
4 -0.979672 -0.990441 -0.942469  ... -0.699205  0.123320  0.122542  0.693578   

        557       558       559       560  Subject  activity  
0 -0.018446 -0.841247  0.179941 -0.058627        1         5  
1  0.703511 -0.844788  0.180289 -0.054317        1         5  
2  0.808529 -0.848933  0.180637 -0.049118        1         5  
3 -0.485366 -0.848649  0.181935 -0.047663        1         5  
4 -0.615971 -0.847865  0.185151 -0.043892        1         5  

[5 rows x 563 columns]

In [11]:
test_df= pd.read_csv('test.csv')

In [12]:
test_df.head()

0         1         2         3         4         5         6  \
0  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   
1  0.286027 -0.013163 -0.119083 -0.975415 -0.967458 -0.944958 -0.986799   
2  0.275485 -0.026050 -0.118152 -0.993819 -0.969926 -0.962748 -0.994403   
3  0.270298 -0.032614 -0.117520 -0.994743 -0.973268 -0.967091 -0.995274   
4  0.274833 -0.027848 -0.129527 -0.993852 -0.967445 -0.978295 -0.994111   

          7         8         9  ...       553       554       555       556  \
0 -0.925249 -0.674302 -0.894088  ... -0.705974  0.006462  0.162920 -0.825886   
1 -0.968401 -0.945823 -0.894088  ... -0.594944 -0.083495  0.017500 -0.434375   
2 -0.970735 -0.963483 -0.939260  ... -0.640736 -0.034956  0.202302  0.064103   
3 -0.974471 -0.968897 -0.938610  ... -0.736124 -0.017067  0.154438  0.340134   
4 -0.965953 -0.977346 -0.938610  ... -0.846595 -0.002223 -0.040046  0.736715   

        557       558       559       560  Subject  activity  
0  0.271151 -0.720009  0.276801 -0.057978        2         5  
1  0.920593 -0.698091  0.281343 -0.083898        2         5  
2  0.145068 -0.702771  0.280083 -0.079346        2         5  
3  0.296407 -0.698954  0.284114 -0.077108        2         5  
4 -0.118545 -0.692245  0.290722 -0.073857        2         5  

[5 rows x 563 columns]

In [13]:
from sklearn.model_selection import train_test_split
x=train_df.drop('activity',axis=1)
y=train_df['activity']
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size= 0.30)


In [14]:
X_train

0         1         2         3         4         5         6  \
5651  0.264015  0.038842 -0.120877 -0.232794 -0.235580 -0.459693 -0.351933   
4451  0.306381 -0.018493 -0.128582 -0.921448 -0.908650 -0.879550 -0.929979   
622   0.267734 -0.039672 -0.123917 -0.398102 -0.067394 -0.390995 -0.450036   
7062  0.277832 -0.017769 -0.110410 -0.996902 -0.994731 -0.995521 -0.997324   
1072  0.275615 -0.016096 -0.104630 -0.984219 -0.978795 -0.989709 -0.985125   
...        ...       ...       ...       ...       ...       ...       ...   
749   0.279313 -0.018870 -0.105402 -0.993513 -0.984998 -0.989872 -0.993811   
7114  0.278175 -0.011995 -0.137810 -0.226034 -0.239538 -0.320031 -0.314143   
6675  0.275663 -0.016558 -0.107614 -0.994267 -0.980993 -0.985600 -0.995123   
3394  0.199118  0.041223 -0.120062 -0.180991 -0.260253 -0.113017 -0.225105   
361   0.263770 -0.021187 -0.081685 -0.930288 -0.851998 -0.920506 -0.939907   

             7         8         9  ...       552       553       554  \
5651 -0.241603 -0.463831 -0.031382  ... -0.160632 -0.588660  0.211543   
4451 -0.909598 -0.873452 -0.816578  ... -0.077114 -0.484494  0.331370   
622  -0.072338 -0.371033 -0.111432  ...  0.530446  0.365402 -0.069481   
7062 -0.995122 -0.995217 -0.938763  ... -0.696020 -0.901786  0.712904   
1072 -0.979665 -0.989117 -0.925231  ... -0.585568 -0.842904 -0.232298   
...        ...       ...       ...  ...       ...       ...       ...   
749  -0.986611 -0.990899 -0.938709  ... -0.400771 -0.814584  0.338383   
7114 -0.228767 -0.262440  0.400349  ... -0.113030 -0.524219  0.009600   
6675 -0.983223 -0.985292 -0.936769  ... -0.302468 -0.612226  0.312489   
3394 -0.286955 -0.068529 -0.058097  ... -0.636321 -0.864901  0.360055   
361  -0.859458 -0.917505 -0.869579  ... -0.525944 -0.845093  0.087209   

           555       556       557       558       559       560  Subject  
5651 -0.781484  0.766219  0.468048 -0.771033  0.256378  0.017798       26  
4451 -0.219954 -0.524127  0.758900  0.468362 -0.468291 -0.529275       22  
622   0.336357  0.486352 -0.348553 -0.719869  0.257486  0.130601        3  
7062  0.328963  0.050523 -0.194982  0.415036 -0.486136 -0.499386       30  
1072  0.163076  0.123416  0.699386  0.599553 -0.979525  0.043860        6  
...        ...       ...       ...       ...       ...       ...      ...  
749   0.690425 -0.179640 -0.172370  0.631909 -0.840509 -0.150759        5  
7114  0.722612  0.921425  0.163260 -0.859623  0.194148  0.032854       30  
6675  0.057043 -0.307446 -0.713738 -0.659901 -0.037938 -0.182402       29  
3394  0.512898  0.952441  0.311913 -0.552304  0.215513  0.313384       17  
361  -0.413666  0.933008  0.506561 -0.716900  0.276004 -0.065339        3  

[5146 rows x 562 columns]

In [15]:
X_test

0         1         2         3         4         5         6  \
2252  0.277688 -0.010204 -0.093866 -0.988746 -0.958301 -0.944715 -0.990084   
910   0.264269 -0.015356 -0.109646 -0.985357 -0.981168 -0.969412 -0.986388   
1468  0.062631 -0.062839 -0.334600 -0.170677 -0.365133  0.099516 -0.210948   
5854  0.254117 -0.032885 -0.114063  0.319201  0.190407 -0.087869  0.253059   
233   0.275549 -0.014115 -0.108999 -0.997599 -0.985257 -0.995653 -0.997931   
...        ...       ...       ...       ...       ...       ...       ...   
4984  0.240427 -0.057314 -0.207500 -0.363554 -0.309795 -0.009082 -0.378938   
6988  0.269636 -0.029252 -0.126291 -0.977195 -0.855117 -0.937536 -0.981686   
5343  0.288712 -0.026788 -0.091126 -0.945106 -0.909591 -0.921883 -0.956524   
6792  0.412214 -0.049787 -0.171298  0.005875 -0.159013  0.079105 -0.091228   
5595  0.269334 -0.018941 -0.142856 -0.379726 -0.257150 -0.310673 -0.424854   

             7         8         9  ...       552       553       554  \
2252 -0.954226 -0.947678 -0.929580  ... -0.274639 -0.594821 -0.038191   
910  -0.983498 -0.964832 -0.930306  ... -0.181958 -0.621825 -0.161157   
1468 -0.404148  0.103444  0.092464  ...  0.207791 -0.149349  0.041817   
5854  0.092310 -0.111886  0.612220  ... -0.454943 -0.822003  0.313863   
233  -0.982909 -0.995529 -0.944120  ... -0.613866 -0.806363  0.102840   
...        ...       ...       ...  ...       ...       ...       ...   
4984 -0.356056 -0.034777 -0.109856  ... -0.233486 -0.655966  0.003321   
6988 -0.845338 -0.933870 -0.903835  ...  0.016999 -0.409799  0.016814   
5343 -0.918665 -0.921316 -0.918146  ...  0.181531 -0.124500  0.043968   
6792 -0.160152  0.027662  0.328099  ... -0.475866 -0.836211 -0.545735   
5595 -0.300854 -0.338936  0.062942  ...  0.486530  0.246215  0.177202   

           555       556       557       558       559       560  Subject  
2252  0.136339  0.167673 -0.303669 -0.912914  0.101725 -0.046527       14  
910   0.386125  0.000866  0.164456  0.584031 -0.811886 -0.182860        5  
1468  0.121594  0.967539 -0.117698 -0.536821  0.207750  0.328708        7  
5854 -0.836786  0.925931  0.028904 -0.836401  0.143155 -0.091134       26  
233  -0.797698 -0.276761  0.015448 -0.734412  0.004690 -0.148719        1  
...        ...       ...       ...       ...       ...       ...      ...  
4984  0.335036  0.937601 -0.201878 -0.800866  0.177783  0.134977       23  
6988 -0.172873 -0.453204 -0.661587 -0.821403  0.210314  0.071393       30  
5343 -0.089714  0.717292  0.180690  0.438821 -0.559243 -0.428118       25  
6792  0.347236 -0.905622  0.091228 -0.805715  0.227345 -0.020070       29  
5595  0.669642  0.545369  0.636703 -0.846901  0.156140 -0.075041       26  

[2206 rows x 562 columns]

In [16]:
X_train.shape, X_test.shape

((5146, 562), (2206, 562))

In [17]:
y_train.shape, y_test.shape

((5146,), (2206,))

In [18]:
from sklearn.linear_model import LogisticRegression
model= LogisticRegression(max_iter=5000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=5000)

In [19]:
model.score(X_train,y_train)

0.9935872522347454

In [20]:
pred=model.predict(X_test)

In [21]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, pred)

0.9823209428830463

In [22]:
from sklearn.metrics import classification_report
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           1       1.00      1.00      1.00       359
           2       1.00      1.00      1.00       325
           3       1.00      1.00      1.00       300
           4       0.94      0.97      0.95       387
           5       0.97      0.94      0.95       427
           6       1.00      1.00      1.00       408

    accuracy                           0.98      2206
   macro avg       0.98      0.98      0.98      2206
weighted avg       0.98      0.98      0.98      2206



In [23]:
from sklearn.metrics import plot_confusion_matrix, confusion_matrix
confusion_matrix(y_test, pred)

array([[359,   0,   0,   0,   0,   0],
       [  0, 325,   0,   0,   0,   0],
       [  0,   1, 299,   0,   0,   0],
       [  0,   0,   0, 375,  12,   0],
       [  0,   0,   0,  26, 401,   0],
       [  0,   0,   0,   0,   0, 408]], dtype=int64)

In [24]:
from sklearn.metrics import f1_score
f1_score(y_test, pred, average='macro')

0.9838888345193338

### Decision Tree

In [25]:
# plot_confusion_matrix(model,X_test,y_test)

In [26]:
from sklearn.tree import DecisionTreeClassifier
classifier=DecisionTreeClassifier()

In [27]:
classifier.fit(X_train,y_train)

DecisionTreeClassifier()

In [28]:
y_pred=classifier.predict(X_test)

In [29]:
y_pred

array([4, 6, 2, ..., 6, 2, 1], dtype=int64)

In [30]:
from sklearn.metrics import r2_score
score=r2_score(y_pred,y_test)

In [31]:
score

0.9705814170501029

In [32]:
parameter={
    'criterion':['gini','entropy'],
  'splitter':['best','random'],
  'max_features':['auto', 'sqrt', 'log2'],
    'max_depth': [3,6,9,12,15,20]
    
}
classifier = DecisionTreeClassifier()

In [33]:
from sklearn.model_selection import GridSearchCV
classifierCv=GridSearchCV(classifier,param_grid=parameter,cv=5,scoring='neg_mean_squared_error')

In [34]:
classifierCv.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 6, 9, 12, 15, 20],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'splitter': ['best', 'random']},
             scoring='neg_mean_squared_error')

In [35]:
classifierCv.best_params_

{'criterion': 'gini',
 'max_depth': 12,
 'max_features': 'sqrt',
 'splitter': 'best'}

In [36]:
y_pred=classifierCv.predict(X_test)

In [37]:
r2_score(y_pred,y_test)

0.9508832300691483

In [38]:
confusion_matrix(y_test,y_pred)

array([[329,  25,   4,   0,   0,   1],
       [ 20, 291,  14,   0,   0,   0],
       [ 10,  18, 272,   0,   0,   0],
       [  0,   0,   0, 323,  56,   8],
       [  0,   0,   0,  45, 372,  10],
       [  0,   0,   1,   4,   3, 400]], dtype=int64)

### Gradient Boosting

In [39]:
from sklearn.ensemble import GradientBoostingClassifier

In [40]:
# from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import GridSearchCV
classifiercv=GridSearchCV(classifier,param_grid=parameter,cv=5,scoring='accuracy')

In [41]:
# param_grid = {"n_estimators":[1,5,10,20,40,100],'max_depth':[3,4,5,6]}
classifiercv.fit(X_train,y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 6, 9, 12, 15, 20],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'splitter': ['best', 'random']},
             scoring='accuracy')

In [42]:
# gb_model = GradientBoostingClassifier

In [43]:
# grid = GridSearchCV(gb_model,param_grid)

In [44]:
# grid.fit(X_train,y_train)

In [45]:
from sklearn.metrics import classification_report,plot_confusion_matrix,accuracy_score

In [46]:
predication = classifiercv.predict(X_test)

In [47]:
predication

array([4, 6, 2, ..., 6, 2, 1], dtype=int64)

In [48]:
print(classification_report(y_test,predication))

              precision    recall  f1-score   support

           1       0.92      0.92      0.92       359
           2       0.91      0.90      0.90       325
           3       0.93      0.94      0.93       300
           4       0.83      0.85      0.84       387
           5       0.85      0.85      0.85       427
           6       1.00      0.99      0.99       408

    accuracy                           0.91      2206
   macro avg       0.91      0.91      0.91      2206
weighted avg       0.91      0.91      0.91      2206



### LSTM

In [49]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import numpy as np

In [50]:
train_df = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt", delim_whitespace = True,names=list(range(0,561)))

train_df['Subject'] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\subject_train.txt", header = None, squeeze = True)

train_df["activity"] = pd.read_csv(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\y_train.txt", header = None, squeeze = True)
# C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train

In [51]:
features = list()
with open(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [52]:
features = list()
with open(r"C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset/features.txt") as f:
    for line in f:
        features.append(line.split()[1])

In [53]:
train_df.to_csv('train.csv',index_label=False)

In [54]:
test_df.to_csv('test.csv',index_label=False)

In [55]:
train_df= pd.read_csv('train.csv')

In [56]:
train_df.head()

0         1         2         3         4         5         6  \
0  0.288585 -0.020294 -0.132905 -0.995279 -0.983111 -0.913526 -0.995112   
1  0.278419 -0.016411 -0.123520 -0.998245 -0.975300 -0.960322 -0.998807   
2  0.279653 -0.019467 -0.113462 -0.995380 -0.967187 -0.978944 -0.996520   
3  0.279174 -0.026201 -0.123283 -0.996091 -0.983403 -0.990675 -0.997099   
4  0.276629 -0.016570 -0.115362 -0.998139 -0.980817 -0.990482 -0.998321   

          7         8         9  ...       553       554       555       556  \
0 -0.983185 -0.923527 -0.934724  ... -0.710304 -0.112754  0.030400 -0.464761   
1 -0.974914 -0.957686 -0.943068  ... -0.861499  0.053477 -0.007435 -0.732626   
2 -0.963668 -0.977469 -0.938692  ... -0.760104 -0.118559  0.177899  0.100699   
3 -0.982750 -0.989302 -0.938692  ... -0.482845 -0.036788 -0.012892  0.640011   
4 -0.979672 -0.990441 -0.942469  ... -0.699205  0.123320  0.122542  0.693578   

        557       558       559       560  Subject  activity  
0 -0.018446 -0.841247  0.179941 -0.058627        1         5  
1  0.703511 -0.844788  0.180289 -0.054317        1         5  
2  0.808529 -0.848933  0.180637 -0.049118        1         5  
3 -0.485366 -0.848649  0.181935 -0.047663        1         5  
4 -0.615971 -0.847865  0.185151 -0.043892        1         5  

[5 rows x 563 columns]

In [57]:
test_df= pd.read_csv('test.csv')

In [58]:
test_df.head()

0         1         2         3         4         5         6  \
0  0.257178 -0.023285 -0.014654 -0.938404 -0.920091 -0.667683 -0.952501   
1  0.286027 -0.013163 -0.119083 -0.975415 -0.967458 -0.944958 -0.986799   
2  0.275485 -0.026050 -0.118152 -0.993819 -0.969926 -0.962748 -0.994403   
3  0.270298 -0.032614 -0.117520 -0.994743 -0.973268 -0.967091 -0.995274   
4  0.274833 -0.027848 -0.129527 -0.993852 -0.967445 -0.978295 -0.994111   

          7         8         9  ...       553       554       555       556  \
0 -0.925249 -0.674302 -0.894088  ... -0.705974  0.006462  0.162920 -0.825886   
1 -0.968401 -0.945823 -0.894088  ... -0.594944 -0.083495  0.017500 -0.434375   
2 -0.970735 -0.963483 -0.939260  ... -0.640736 -0.034956  0.202302  0.064103   
3 -0.974471 -0.968897 -0.938610  ... -0.736124 -0.017067  0.154438  0.340134   
4 -0.965953 -0.977346 -0.938610  ... -0.846595 -0.002223 -0.040046  0.736715   

        557       558       559       560  Subject  activity  
0  0.271151 -0.720009  0.276801 -0.057978        2         5  
1  0.920593 -0.698091  0.281343 -0.083898        2         5  
2  0.145068 -0.702771  0.280083 -0.079346        2         5  
3  0.296407 -0.698954  0.284114 -0.077108        2         5  
4 -0.118545 -0.692245  0.290722 -0.073857        2         5  

[5 rows x 563 columns]

In [59]:
from sklearn.model_selection import train_test_split
x=train_df.drop('activity',axis=1)
y=train_df['activity']
X_train, X_test, y_train, y_test = train_test_split(x,y,test_size= 0.30)

In [60]:
import tensorflow as tf

In [61]:
# univariate lstm example
import numpy as np
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Flatten

In [62]:
from keras import backend as K
from keras.models import Sequential
from keras.layers import LSTM
from keras.layers.core import Dense, Dropout
from keras.layers import BatchNormalization
from keras.regularizers import L1L2

In [63]:
# Activities are the class labels
# It is a 6 class classification
ACTIVITIES = {
    0: 'WALKING',
    1: 'WALKING_UPSTAIRS',
    2: 'WALKING_DOWNSTAIRS',
    3: 'SITTING',
    4: 'STANDING',
    5: 'LAYING',
}

In [64]:
import matplotlib.pyplot as plt
import seaborn as sns

# function to print the confusion matrix

def confusion_matrix(Y_true, Y_pred):
    
    Y_true = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_true, axis=1)])
    Y_pred = pd.Series([ACTIVITIES[y] for y in np.argmax(Y_pred, axis=1)])

    return pd.crosstab(Y_true, Y_pred, rownames=['True'], colnames=['Pred'])

  
    plt.show() 

In [65]:
SIGNALS = ["body_acc_x","body_acc_y","body_acc_z","body_gyro_x","body_gyro_y","body_gyro_z","total_acc_x","total_acc_y","total_acc_z"]

In [66]:
# function to read the data from csv file
def _read_csv(filename):
    return pd.read_csv(filename, delim_whitespace=True, header=None)

# function to load the load
def load_signals(subset):
    signals_data = []

    for signal in SIGNALS:
        filename = f'C:/Users/shant/Downloads/UCI HAR Dataset/UCI HAR Dataset/{subset}/Inertial Signals/{signal}_{subset}.txt'
        
        signals_data.append(
            _read_csv(filename).to_numpy()
        ) 

   
    return np.transpose(signals_data, (1, 2, 0))


# C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset
# C:\Users\shant\Downloads\UCI HAR Dataset\UCI HAR Dataset\train\X_train.txt

In [67]:
def load_y(subset):
    filename = f'C:/Users/shant/Downloads/UCI HAR Dataset/UCI HAR Dataset/{subset}/y_{subset}.txt'
    y = _read_csv(filename)[0]

    return pd.get_dummies(y).to_numpy()

In [68]:
def load_data():
    X_train, X_test = load_signals('train'), load_signals('test')
    y_train, y_test = load_y('train'), load_y('test')

    return X_train, X_test, y_train, y_test

In [69]:
# Importing tensorflow
np.random.seed(42)
import tensorflow as tf
tf.random.set_seed(42)

In [70]:
# Initializing parameters
epochs = 50
batch_size = 16
n_hidden = 32

In [71]:
#function to count the number of classes
def _count_classes(y):
    return len(set([tuple(category) for category in y]))

In [72]:
X_train, X_test, Y_train, Y_test = load_data()

In [73]:
timesteps = len(X_train[0])
input_dim = len(X_train[0][0])
n_classes = _count_classes(Y_train)

print(timesteps)
print(input_dim)
print(len(X_train))

128
9
7352


In [74]:
# Initiliazing the sequential model
model = Sequential()
# Configuring the parameters
model.add(LSTM(n_hidden, input_shape=(timesteps, input_dim)))
# Adding a dropout layer
model.add(Dropout(0.5))
# Adding a dense output layer with sigmoid activation
model.add(Dense(n_classes, activation='sigmoid'))
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm (LSTM)                 (None, 32)                5376      
                                                                 
 dropout (Dropout)           (None, 32)                0         
                                                                 
 dense (Dense)               (None, 6)                 198       
                                                                 
Total params: 5,574
Trainable params: 5,574
Non-trainable params: 0
_________________________________________________________________


In [75]:
# Compiling the model
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

In [76]:
# Training the model
model.fit(X_train, Y_train, batch_size=batch_size, validation_data=(X_test, Y_test),epochs=epochs)

Epoch 1/50
460/460 [==============================] - 17s 33ms/step - loss: 1.2511 - accuracy: 0.4748 - val_loss: 1.1105 - val_accuracy: 0.5399
Epoch 2/50
460/460 [==============================] - 14s 31ms/step - loss: 0.8770 - accuracy: 0.6502 - val_loss: 1.1187 - val_accuracy: 0.6274
Epoch 3/50
460/460 [==============================] - 16s 34ms/step - loss: 0.6304 - accuracy: 0.7489 - val_loss: 0.7466 - val_accuracy: 0.7329
Epoch 4/50
460/460 [==============================] - 16s 34ms/step - loss: 0.5232 - accuracy: 0.8022 - val_loss: 0.6864 - val_accuracy: 0.8035
Epoch 5/50
460/460 [==============================] - 14s 31ms/step - loss: 0.4583 - accuracy: 0.8530 - val_loss: 0.6394 - val_accuracy: 0.8018
Epoch 6/50
460/460 [==============================] - 14s 31ms/step - loss: 0.3965 - accuracy: 0.8742 - val_loss: 0.6306 - val_accuracy: 0.8314
Epoch 7/50
460/460 [==============================] - 14s 30ms/step - loss: 0.3380 - accuracy: 0.8930 - val_loss: 0.6227 - val_accuracy:

In [77]:
(Y_test, model.predict(X_test))

93/93 [==============================] - 1s 7ms/step


(array([[0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        ...,
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0]], dtype=uint8),
 array([[1.48081139e-03, 6.76601660e-04, 1.12245769e-04, 7.18471467e-01,
         9.96061206e-01, 2.83957412e-03],
        [1.57628942e-03, 7.88853155e-04, 1.20411161e-04, 7.17142344e-01,
         9.95490551e-01, 2.73726624e-03],
        [1.54523249e-03, 7.74537737e-04, 1.16243784e-04, 7.19521940e-01,
         9.95529115e-01, 2.75537302e-03],
        ...,
        [7.53785491e-01, 9.99999702e-01, 7.72234440e-01, 2.27266047e-02,
         1.64474137e-02, 5.55884391e-02],
        [7.04586089e-01, 9.99999821e-01, 8.08141291e-01, 2.61969510e-02,
         1.73719935e-02, 2.51788855e-01],
        [7.09573567e-01, 9.99999583e-01, 7.74231970e-01, 3.22261453e-02,
         3.06887906e-02, 6.78987086e-01]], dtype=float32))

In [78]:
score = model.evaluate(X_test, Y_test)

print("\n   cat_crossentropy  ||   accuracy ")
print("  ____________________________________")
print(score)

93/93 [==============================] - 1s 7ms/step - loss: 0.2972 - accuracy: 0.9186

   cat_crossentropy  ||   accuracy 
  ____________________________________
[0.2972451150417328, 0.9185612201690674]


In [79]:
# Initializing parameters
n_epochs = 50
n_batch = 16
n_classes = _count_classes(Y_train)

# Bias regularizer value - we will use elasticnet
reg = L1L2(0.01, 0.01)

In [80]:
# Model execution
model = Sequential()
model.add(LSTM(48, input_shape=(timesteps, input_dim), return_sequences=True,bias_regularizer=reg ))
model.add(BatchNormalization())
model.add(Dropout(0.50))
model.add(LSTM(32))
model.add(Dropout(0.50))
model.add(Dense(n_classes, activation='sigmoid'))
print("Model Summary: ")
model.summary()

Model Summary: 
Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm_1 (LSTM)               (None, 128, 48)           11136     
                                                                 
 batch_normalization (BatchN  (None, 128, 48)          192       
 ormalization)                                                   
                                                                 
 dropout_1 (Dropout)         (None, 128, 48)           0         
                                                                 
 lstm_2 (LSTM)               (None, 32)                10368     
                                                                 
 dropout_2 (Dropout)         (None, 32)                0         
                                                                 
 dense_1 (Dense)             (None, 6)                 198       
                                      

In [81]:
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [82]:
# Training the model
model.fit(X_train, Y_train, batch_size=n_batch, validation_data=(X_test, Y_test), epochs=n_epochs)

Epoch 1/50
460/460 [==============================] - 28s 54ms/step - loss: 1.5326 - accuracy: 0.6925 - val_loss: 1.0566 - val_accuracy: 0.8079
Epoch 2/50
460/460 [==============================] - 24s 52ms/step - loss: 0.7115 - accuracy: 0.8796 - val_loss: 0.7059 - val_accuracy: 0.7815
Epoch 3/50
460/460 [==============================] - 28s 61ms/step - loss: 0.4221 - accuracy: 0.9083 - val_loss: 0.4387 - val_accuracy: 0.8751
Epoch 4/50
460/460 [==============================] - 28s 61ms/step - loss: 0.2957 - accuracy: 0.9117 - val_loss: 0.2786 - val_accuracy: 0.9084
Epoch 5/50
460/460 [==============================] - 29s 63ms/step - loss: 0.2274 - accuracy: 0.9264 - val_loss: 0.2441 - val_accuracy: 0.9084
Epoch 6/50
460/460 [==============================] - 27s 58ms/step - loss: 0.2150 - accuracy: 0.9286 - val_loss: 0.3152 - val_accuracy: 0.8962
Epoch 7/50
460/460 [==============================] - 26s 57ms/step - loss: 0.2162 - accuracy: 0.9261 - val_loss: 0.2198 - val_accuracy:

In [83]:
# Confusion Matrix
#new_confusion_matrix(Y_test, model.predict(X_test))
Y_test, model.predict(X_test)

93/93 [==============================] - 2s 16ms/step


(array([[0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        [0, 0, 0, 0, 1, 0],
        ...,
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0],
        [0, 1, 0, 0, 0, 0]], dtype=uint8),
 array([[7.2816839e-03, 2.8329431e-03, 1.3407016e-04, 9.7422141e-01,
         9.9908018e-01, 3.9036882e-03],
        [7.0112711e-03, 2.7990618e-03, 1.3515892e-04, 9.7449988e-01,
         9.9908400e-01, 3.7507284e-03],
        [7.1442565e-03, 2.6248880e-03, 1.2908292e-04, 9.7535944e-01,
         9.9912775e-01, 3.6180462e-03],
        ...,
        [3.3867845e-01, 9.9999678e-01, 2.8936481e-01, 8.2610012e-04,
         1.5068665e-01, 3.4840617e-02],
        [6.6389608e-01, 9.9999768e-01, 2.0058317e-01, 8.7830058e-04,
         8.2855366e-02, 5.9065875e-02],
        [5.5981696e-01, 9.9999774e-01, 1.6682929e-01, 7.8353658e-04,
         1.0108251e-01, 7.1388960e-02]], dtype=float32))

In [84]:
score = model.evaluate(X_test, Y_test)

print("\n   cat_crossentropy  ||   accuracy ")
print("  ____________________________________")
print(score)

93/93 [==============================] - 1s 16ms/step - loss: 0.2867 - accuracy: 0.9220

   cat_crossentropy  ||   accuracy 
  ____________________________________
[0.28665891289711, 0.9219545125961304]


### Accuracy

#### Logistic Regression:- 98.38%

#### Decision Tree:-95.08%

#### Gradient Boosting:-91%

#### LSTM:-92.19%